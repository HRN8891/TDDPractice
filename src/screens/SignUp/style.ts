import {StyleSheet} from 'react-native';
import {scale, verticalScale} from 'react-native-size-matters';
import {Colors} from '../../utilities/Colors';

const styles = StyleSheet.create({
  screen: {
    paddingTop: 25,
    paddingBottom: verticalScale(24),
    paddingHorizontal: scale(16),
    flex: 1,
    backgroundColor: Colors.grey4,
  },
  all: {
    marginTop: verticalScale(24),
    flex: 1,
    justifyContent: 'center',
  },
  crossButton: {color: Colors.grey1, margin: 16},
});

export default styles;
