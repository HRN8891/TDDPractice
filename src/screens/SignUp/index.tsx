import React, {useState} from 'react';
import {SafeAreaView, Alert, TouchableOpacity} from 'react-native';
import {KeyboardAwareScrollView} from '@codler/react-native-keyboard-aware-scroll-view';

import style from './style';
import {strings} from '../../locales/i18n';
import CTextBox from '../../components/CTextBox';
import CButton from '../../components/CButton';
import {setSecureStorageItem, storageKeys} from '../../utilities/storageUtils';
import {validateEmail} from '../../utilities/misc';
import {goBack} from '../../navigation/root';
import {useSignUpMutation} from 'store/auth/service';
import {saveUserDetails} from 'store/auth/slice';
import {useAppDispatch} from 'store';

const {alert} = Alert;

function SignUp() {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [signUp] = useSignUpMutation();
  const dispatch = useAppDispatch();

  const handleSignUp = async () => {
    try {
      const trimmedEmail = email.trim();
      const trimmedPassword = password.trim();
      const trimmedConfirmPassword = confirmPassword.trim();

      if (!trimmedEmail) {
        return alert('Email is required');
      }
      if (!validateEmail(trimmedEmail)) {
        return alert('Please enter valid Email');
      }
      if (!trimmedPassword) {
        return alert('Password is required');
      }
      if (!trimmedConfirmPassword) {
        return alert('Confirm Password is required');
      }
      if (trimmedPassword !== trimmedConfirmPassword) {
        return alert('Password & Confirm Password do not match');
      }

      const data = await signUp({
        email: trimmedEmail,
        password: trimmedPassword,
      }).unwrap();
      await setSecureStorageItem(storageKeys.userData, data);
      dispatch(saveUserDetails(data));
    } catch (error) {
      alert(strings('common.something_went_wrong'));
    }
  };

  function renderBackButton() {
    return <TouchableOpacity onPress={goBack}>Add back button icon here</TouchableOpacity>;
  }

  return (
    <SafeAreaView testID="Sigunup Component" style={style.screen}>
      {renderBackButton()}
      <KeyboardAwareScrollView
        contentContainerStyle={style.all}
        keyboardShouldPersistTaps={'handled'}>
        <CTextBox
          placeholder={strings('signUp.email')}
          value={email}
          onChangeText={setEmail}
          keyboardType="email-address"
          testID="Email"
        />
        <CTextBox
          placeholder={strings('signUp.password')}
          value={password}
          onChangeText={setPassword}
          hidePassword={true}
          testID="Password"
        />
        <CTextBox
          placeholder={strings('signUp.confirmPassword')}
          value={confirmPassword}
          onChangeText={setConfirmPassword}
          hidePassword={true}
          returnKeyType={'done'}
          testID="ConfrimPassword"
        />
        <CButton testID="Signup Button" onPress={handleSignUp} text={strings('signUp.signUp')} />
      </KeyboardAwareScrollView>
    </SafeAreaView>
  );
}

export default SignUp;
