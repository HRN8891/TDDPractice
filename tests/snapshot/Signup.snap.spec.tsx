import React from 'react';
import SignUp from '../../src/screens/SignUp';
import {fireEvent, render} from '@testing-library/react-native';
import ErrorBoundary from '../../src/components/ErrorBoundary';
import {Provider} from 'react-redux';
import {View, Alert} from 'react-native';

jest.mock('../../src/components/ErrorBoundary', () => jest.fn().mockReturnValue(null));
jest.mock('react-redux', () => {
  return {
    ...jest.requireActual<object>('react-redux'),
    Provider: jest.fn(),
    useSelector: jest.fn(fn => fn()),
    useDispatch: () => jest.fn().mockImplementation(() => Promise.resolve()),
  };
});

jest.mock('../../src/navigation/root', () => 'navigationRef');

describe('SignUp screen', () => {
  it('SignUp screen', () => {
    (Provider as jest.Mock).mockImplementationOnce(({children}) => children);
    (ErrorBoundary as jest.Mock).mockReturnValue(<View />);
    const tree = render(<SignUp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('SignUp screen Test', () => {
  test('Should render correctly', () => {
    const wrapper = render(<SignUp />);
    wrapper.getByTestId('Sigunup Component');
  });

  test('Render Email', () => {
    const wrapper = render(<SignUp />);
    wrapper.getByTestId('Email');
  });

  test('Render Password', () => {
    const wrapper = render(<SignUp />);
    wrapper.getByTestId('Password');
  });

  test('Render Confirm Password', () => {
    const wrapper = render(<SignUp />);
    wrapper.getByTestId('ConfrimPassword');
  });

  test('Render Email Error', () => {
    const wrapper = render(<SignUp />);
    fireEvent.press(wrapper.getByRole('button', {name: 'Sign Up'}));
    // wrapper.getByTestId('Error');
    // expect(await wrapper.findByRole('alert')).toHaveTextContent('Incorrect username or password');
    expect(Alert.alert).toHaveBeenCalledWith('Email is required'); // passes
  });

  test('Render Password Error', () => {
    const wrapper = render(<SignUp />);
    // const input = wrapper.getByTestId('Email');
    // fireEvent.changeText(input, 'admin@gmail.com');
    // expect(input.props.value).toBe('admin@gmail.com');
    fireEvent.changeText(wrapper.getByTestId('Email'), 'admin@gmail.com');
    fireEvent.press(wrapper.getByRole('button', {name: 'Sign Up'}));
    expect(Alert.alert).toHaveBeenCalledWith('Password is required'); // passes
  });

  test('Password is required', () => {
    const wrapper = render(<SignUp />);
    fireEvent.changeText(wrapper.getByTestId('Email'), 'admin@mail.com');
    fireEvent.press(wrapper.getByRole('button', {name: 'Sign Up'}));
    expect(Alert.alert).toHaveBeenCalledWith('Password is required'); // passes
  });

  test('Password does not match', () => {
    const wrapper = render(<SignUp />);
    fireEvent.changeText(wrapper.getByTestId('Email'), 'admin@gmail.com');
    fireEvent.changeText(wrapper.getByTestId('Password'), 'Admin123');
    fireEvent.changeText(wrapper.getByTestId('ConfrimPassword'), 'Admin12377');
    fireEvent.press(wrapper.getByRole('button', {name: 'Sign Up'}));
    expect(Alert.alert).toHaveBeenCalledWith('Password & Confirm Password do not match'); // passes
  });

  test('Render Confirm Password Error', () => {
    const wrapper = render(<SignUp />);
    fireEvent.changeText(wrapper.getByTestId('Email'), 'admin@gmail.com');
    fireEvent.changeText(wrapper.getByTestId('Password'), 'Admin123');
    fireEvent.press(wrapper.getByRole('button', {name: 'Sign Up'}));
    expect(Alert.alert).toHaveBeenCalledWith('Confirm Password is required'); // passes
  });
});
