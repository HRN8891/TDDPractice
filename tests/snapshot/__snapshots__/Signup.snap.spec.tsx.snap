// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp screen SignUp screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(245, 245, 245)",
      "flex": 1,
      "paddingBottom": 47.08235294117647,
      "paddingHorizontal": 34.285714285714285,
      "paddingTop": 25,
    }
  }
  testID="Sigunup Component"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    Add back button icon here
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 25.714285714285715,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      caretHidden={false}
      keyboardType="email-address"
      onChangeText={[Function]}
      selectionColor="#189E66"
      style={
        Array [
          Object {
            "backgroundColor": "#F1F2F1",
            "borderRadius": 21.428571428571427,
            "fontSize": 32.14285714285714,
            "height": 96.42857142857143,
            "paddingLeft": 10,
            "paddingVertical": 8.571428571428571,
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
      testID="Email"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 25.714285714285715,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      caretHidden={false}
      onChangeText={[Function]}
      secureTextEntry={true}
      selectionColor="#189E66"
      style={
        Array [
          Object {
            "backgroundColor": "#F1F2F1",
            "borderRadius": 21.428571428571427,
            "fontSize": 32.14285714285714,
            "height": 96.42857142857143,
            "paddingLeft": 10,
            "paddingVertical": 8.571428571428571,
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
      testID="Password"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "marginTop": 25.714285714285715,
        },
        undefined,
      ]
    }
  >
    <TextInput
      autoCapitalize="none"
      caretHidden={false}
      onChangeText={[Function]}
      returnKeyType="done"
      secureTextEntry={true}
      selectionColor="#189E66"
      style={
        Array [
          Object {
            "backgroundColor": "#F1F2F1",
            "borderRadius": 21.428571428571427,
            "fontSize": 32.14285714285714,
            "height": 96.42857142857143,
            "paddingLeft": 10,
            "paddingVertical": 8.571428571428571,
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
      testID="ConfrimPassword"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityLabel=""
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#189E66",
          "borderRadius": 21.428571428571427,
          "justifyContent": "center",
          "opacity": 1,
          "paddingVertical": 27.857142857142858,
          "width": "100%",
        }
      }
      testID=""
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 34.285714285714285,
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
</View>
`;
